README.rst
flask-jinjahelpers
==================

Various helpers for Jinja2 templates

CHANGES.rst
Changelog
---------

Here you can see the full list of changes between each Flask-JinjaHelpers release.


0.3.8 (2017-01-21)
^^^^^^^^^^^^^^^^^^

- Add Python 3 support


0.3.7 (2014-10-08)
^^^^^^^^^^^^^^^^^^

- Change render_header_link max_sorted_fields default from 2 to 1


0.3.6 (2014-10-06)
^^^^^^^^^^^^^^^^^^

- Default value of max_sorted_fields in header_sort_url changed from 2 to 1
- Removed deprecated function qp_url_for


0.3.5 (2013-10-09)
^^^^^^^^^^^^^^^^^^

- Removed unused 'view' parameter from render_pagination


0.3.4 (2013-10-09)
^^^^^^^^^^^^^^^^^^

- Added request.view_args support for url_for_current


0.3.3 (2013-08-05)
^^^^^^^^^^^^^^^^^^

- Added callback argument for url_for_current func


0.3.2 (2013-06-14)
^^^^^^^^^^^^^^^^^^

- render_pagination macro now uses url_for_current


0.3.1 (2013-06-14)
^^^^^^^^^^^^^^^^^^

- url_for_current added in init_jinja_env


0.3.0 (2013-06-14)
^^^^^^^^^^^^^^^^^^

- qp_url_for replaced by url_for_current


0.2.4 (2013-06-11)
^^^^^^^^^^^^^^^^^^

- Made sorted_fields parameter of header_sort_url function accept strings


0.2.3 (2013-06-11)
^^^^^^^^^^^^^^^^^^

- Added max_sorted_fields parameter to header sorter


0.2.2 (2013-06-11)
^^^^^^^^^^^^^^^^^^

- Fixed header_sort_url to support consecutive calls


0.2.1 (2013-06-11)
^^^^^^^^^^^^^^^^^^

- Fixed render_header_link


0.2.0 (2013-06-11)
^^^^^^^^^^^^^^^^^^

- Added multi-column sorter


tests/__init__.py
from flask import Flask
from flask_jinjahelpers import init_jinja_env, url_for_current
import pytest


class TemplateHelperTestCase(object):
    def setup_method(self, method):
        class Application(Flask):
            def create_jinja_environment(self):
                rv = super(Application, self).create_jinja_environment()
                return init_jinja_env(rv)

        self.app = Application(__name__)
        self.app.debug = True
        self.setup_views()

        self.client = self.app.test_client()
        self.context = self.app.test_request_context()
        self.context.push()

    def teardown_method(self, method):
        self.context.pop()
        self.context = None
        self.client = None
        self.app = None


@pytest.mark.parametrize(
    ('requested_url', 'expected_url'),
    [
        ('/view1/', '/view1/'),
        ('/view1/?foo=bar', '/view1/?foo=bar'),
        ('/view1/?foo=bar&foo=baz', '/view1/?foo=bar&foo=baz'),
        ('/view2/1', '/view2/1'),
        ('/view2/1?foo=bar', '/view2/1?foo=bar'),
        ('/view2/1?foo=bar&foo=baz', '/view2/1?foo=bar&foo=baz'),
        ('/view3/', '/view3/?foo=baz'),
        ('/view3/?foo=bar', '/view3/?foo=baz')
    ]
)
def test_url_for_current(requested_url, expected_url):
    app = Flask(__name__)
    app.debug = True

    @app.route('/view1/')
    def view1():
        return url_for_current()

    @app.route('/view2/<int:id>')
    def view2(id):
        return url_for_current()

    @app.route('/view3/')
    def view3():
        return url_for_current(foo='baz')

    client = app.test_client()
    response = client.get(requested_url)
    assert response.data == expected_url


tests/test_header_sorter.py
from flask import render_template_string, url_for, request
from flask_jinjahelpers import header_sort_url
from tests import TemplateHelperTestCase


class TestHeaderSorter(TemplateHelperTestCase):
    def setup_views(self):
        @self.app.route('/')
        def index():
            sorted_fields = request.args.getlist('sort')
            return render_template_string(
                """
                {% from "_tablehelpers.html" import render_header_link %}
                {{
                render_header_link('name', 'Name', sorted_fields)
                }}
                """,
                sorted_fields=sorted_fields
            )

    def test_sort_by_empty_string(self):
        assert header_sort_url('') == (
            '/'
        )

    def test_with_consecutive_calls(self):
        assert header_sort_url('age') == (
            '/?sort=age'
        )
        assert header_sort_url('name') == (
            '/?sort=name'
        )

    def test_with_empty_string_in_sorted_fields(self):
        assert header_sort_url('name', ['']) == (
            '/?sort=name'
        )

    def test_with_hyphen_in_sorted_fields(self):
        assert header_sort_url('name', ['-']) == (
            '/?sort=name'
        )

    def test_sorted_fields_supports_string_as_parameter(self):
        assert header_sort_url('age', 'age') == (
            '/?sort=-age'
        )

    def test_sort_by_single_header(self):
        assert header_sort_url('age', ['name'], max_sorted_fields=2) == (
            '/?sort=age&sort=name'
        )
        assert header_sort_url('name') == '/?sort=name'
        assert header_sort_url('name', ['-name']) == '/?sort=name'
        assert header_sort_url('name', ['name']) == '/?sort=-name'

    def test_with_multiple_sorted_fields(self):
        assert header_sort_url(
            'name', ['name', 'age'], max_sorted_fields=2
        ) == (
            '/?sort=-name&sort=age'
        )

        assert header_sort_url(
            'age', ['name', 'age'], max_sorted_fields=2
        ) == (
            '/?sort=-age&sort=name'
        )

        assert header_sort_url(
            'age', ['name', '-age'], max_sorted_fields=2
        ) == (
            '/?sort=age&sort=name'
        )

        assert header_sort_url(
            'age', ['-age', 'name'], max_sorted_fields=2
        ) == (
            '/?sort=age&sort=name'
        )

    def test_max_number_of_sorted_fields(self):
        assert header_sort_url(
            'name', ['name', 'age'], max_sorted_fields=1
        ) == (
            '/?sort=-name'
        )

    def test_render_header_link(self):
        response = self.client.get(url_for('.index', sort=['name']))

        assert  '<th class="table-sort-asc">' in response.data
        assert '<a href="/?sort=-name">Name</a>' in response.data
        assert '</th>' in response.data

        response = self.client.get(url_for('.index', sort=['-name']))
        assert  '<th class="table-sort-desc">' in response.data
        assert '<a href="/?sort=name">Name</a>' in response.data
        assert '</th>' in response.data

        response = self.client.get(url_for('.index', sort=['name', 'age']))
        assert response.data.count('</th>') == 1


tests/test_visible_page_numbers.py
from flask_jinjahelpers import visible_page_numbers


class TestVisiblePageNumbers(object):
    def test_should_calculate_windowed_visible_pages(self):
        visible = visible_page_numbers(page=7, pages=42)
        assert visible == [1, '...', 4, 5, 6, 7, 8, 9, 10, '...', 42]

    def test_should_eliminate_small_gaps(self):
        visible = visible_page_numbers(page=6, pages=11)
        assert visible == [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

    def test_should_adjust_upper_limit_if_lower_is_out_of_bounds(self):
        visible = visible_page_numbers(page=1, pages=42)
        assert visible == [1, 2, 3, 4, 5, 6, 7, '...', 42]

    def test_should_adjust_lower_limit_if_upper_is_out_of_bounds(self):
        visible = visible_page_numbers(page=40, pages=42)
        assert visible == [1, '...', 36, 37, 38, 39, 40, 41, 42]


flask_jinjahelpers/__init__.py
from flask import request, url_for
from jinja2 import ChoiceLoader, PackageLoader
import six


def init_jinja_env(env):
    """
    Adds flask_jinjahelpers templates and some utility function to given jinja
    enviroment

    :param env: jinja environment
    """
    env.loader = ChoiceLoader([
        env.loader,
        PackageLoader(
            'flask_jinjahelpers',
            'templates'
        )
    ])
    env.globals.update(
        visible_page_numbers=visible_page_numbers,
        url_for_current=url_for_current,
        header_sort_url=header_sort_url
    )
    return env


def url_for_current(callback=None, **kwargs):
    """
    Returns the current url endpoint with all query parameters but replaces
    the query parameters given in kwargs

    :param kwargs: dict containing query parameter names as keys
    """
    data = {}
    data.update(request.args)
    data.update(request.view_args)
    data.update(kwargs)

    if callback:
        data = callback(data)

    return url_for(request.endpoint, **data)


def inverse_sorted_field(sort_by, sorted_fields):
    if isinstance(sorted_fields, six.string_types):
        sorted_fields = [sorted_fields]
    elif sorted_fields is None:
        sorted_fields = []

    sorted_fields = [s for s in sorted_fields if s.strip('-')]
    field_names = [s.strip('-') for s in sorted_fields]

    try:
        field_index = field_names.index(sort_by)
    except ValueError:
        sorted_fields.insert(0, sort_by)
    else:
        old_sort_by = sorted_fields[field_index]
        del sorted_fields[field_index]
        if old_sort_by[0] == '-':
            sorted_fields.insert(0, sort_by)
        else:
            sorted_fields.insert(0, '-%s' % sort_by)
    return sorted_fields


def header_sort_url(
    sort_by, sorted_fields=None, max_sorted_fields=1, **kwargs
):
    if not sort_by:
        return url_for_current(**kwargs)

    sorted_fields = inverse_sorted_field(sort_by, sorted_fields)
    sorted_fields = sorted_fields[0:max_sorted_fields]

    if sorted_fields:
        kwargs['sort'] = sorted_fields
    return url_for_current(**kwargs)


def visible_page_numbers(page, pages, inner_window=3, outer_window=0):
    """
    Takes the current page number and total number of pages, and computes
    an array containing the visible page numbers.  At least three pages on
    either side of the current page as well as the first and last pages
    will be included. For example::

        [1] 2 3 4 5 6 7 ... 42
        1 2 3 4 5 [6] 7 ... 42
        1 ... 4 5 6 [7] 8 9 ... 42
        1 ... 36 37 38 39 [40] 41 42

    :param page: current page number
    :param pages: total number of pages
    """
    window_from = page - inner_window
    window_to = page + inner_window

    if window_to > pages:
        window_from -= window_to - pages
        window_to = pages

    if window_from < 1:
        window_to += 1 - window_from
        window_from = 1
        if window_to > pages:
            window_to = pages

    visible = []

    left_gap_start = min(2 + outer_window, window_from)
    left_gap_end = window_from - 1

    for page in range(1, left_gap_start):
        visible.append(page)

    if left_gap_end - left_gap_start > 0:
        visible.append('...')
    elif left_gap_start == left_gap_end:
        visible.append(left_gap_start)

    right_gap_start = min(window_to + 1, pages - outer_window)
    right_gap_end = pages - outer_window - 1

    for page in range(left_gap_end + 1, right_gap_start):
        visible.append(page)

    if right_gap_end - right_gap_start > 0:
        visible.append('...')
    elif right_gap_start == right_gap_end:
        visible.append(right_gap_start)

    for page in range(right_gap_end + 1, pages + 1):
        visible.append(page)

    return visible
